File: Code/Websites/DanpheEMR/wwwroot/DanpheApp/src/app/billing/shared/credit-note-vms.ts
export class CRN_CreditNoteAllInfoVM {
  public IsInvoiceFound: boolean = false;
  public PatientInfo: CRN_PatientInfoVM = null;
  public InvoiceInfo: CRN_InvoiceInfoVM = null;
  public TransactionItems: Array<CRN_InvoiceItemsVM> = [];
  public AlreadyReturnedItms = [];
}


export class CRN_PatientInfoVM {
  public PatientId: number = 0;
  public PatientCode: string = null;
  public ShortName: string = null;
  public DateOfBirth: string = null;
  public Gender: string = null;
  public CountryName: string = null;
  public CountrySubDivisionName: string = null;
  public Address: string = null;
}

export class CRN_InvoiceInfoVM {
  public PatientId: number = 0;
  public BillingTransactionId: number = 0;
  public InvoiceCode: string = null;
  public InvoiceNo: number = 0;
  public PaymentMode: string = null;
  public InvoiceDate: string = null;
  public InvoiceNoFormatted: string = null;
  public SubTotal: number = 0;
  public DiscountAmount: number = 0;
  public TaxTotal: number = 0;
  public TotalAmount: number = 0;
  public BillStatus: string = null;
  public TransactionType: string = null;
  public InvoiceType: string = null;
  public IsInsuranceBilling: boolean = false;
  public InsuranceProviderId: number = null;
  public UserName: string = null;
  public SettlementId: number = 0;
  public CashDiscount: number = 0;
  public SchemeName: string = "";
  public PriceCategoryId: number = 0;
  public PriceCategoryName: string = "";
  public ClaimCode: number = 0;

  public SchemeId: number = null;//Sud:22march'23--For billing new structure.
  public IsCoPayment: boolean = false;
  public OrganizationId: number = null;
}

export class CRN_InvoiceItemsVM {
  public BillingTransactionItemId: number = 0;
  public BillingTransactionId: number = 0;
  public PatientId: number = 0;
  public ServiceDepartmentId: number = 0;
  public ServiceItemId: number = 0;//Sud:16Apr'23--renamed ItemId to ServiceItemId as per New changes
  public ItemCode: string = null;//sud:16Apr'23--added new column
  public ItemName: string = null;
  public Price: number = 0;
  public RemainingQty: number = 0;
  public SubTotal: number = 0;
  public DiscountAmtPerUnit: number = 0;
  public DiscountAmount: number = 0;
  public TaxAmtPerUnit: number = 0;
  public TaxAmount: number = 0;
  public TotalAmtPerUnit: number = 0;
  public TotalAmount: number = 0;
  public DiscountPercent: number = 0;
  public PrescriberId: number = null;
  public PerformerId: number = null;
  public BillStatus: string = null;
  public RequisitionId: number = null;
  public RequisitionDate: string = null;
  public RequestedBy: number = null;
  public PatientVisitId: number = null;
  public BillingPackageId: number = null;
  public CreatedBy: number = 0;
  public CreatedOn: string = null;
  public BillingType: string = null;
  public RequestingDeptId: number = null;
  public VisitType: string = null;
  public PriceCategory: string = null;
  public PriceCategoryId: number = 0;//sud:16Apr'23--New column added in API
  public PatientInsurancePackageId: number = null;
  public IsInsurance: boolean = false;
  public DiscountSchemeId: number = null;
  public LabTypeName: string = null;
  public OrderStatus: string = null;

  //below two properties are only for local use (client side only)
  public ReturnQuantity: number = 0;
  public IsSelected: boolean = false;
  public IsValid: boolean = true;
  public IsReturnRestricted: boolean = false;
  public SrvDeptIntegrationName: string = null;//needed for orderstatus for return restriction. only in client side.
  public CoPayCashAmount: number = 0;
  public CoPayCreditAmount: number = 0;
}

File: Code/Components/DanpheEMR.ServerModel/BillingModels/CreditNoteVMs.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DanpheEMR.ServerModel.BillingModels
{

    public class CRN_PatientInfoVM
    {
        public int PatientId { get; set; }
        public string PatientCode { get; set; }
        public string ShortName { get; set; }
        public DateTime DateOfBirth { get; set; }
        public string Gender { get; set; }
        public string CountryName { get; set; }
        public string CountrySubDivisionName { get; set; }
        public string Address { get; set; }


        public static CRN_PatientInfoVM GetSinglePatientInfoMappedFromDbTable(DataTable patInfo)
        {
            CRN_PatientInfoVM retObj = new CRN_PatientInfoVM();
            if (patInfo != null)
            {
                string strPatData = JsonConvert.SerializeObject(patInfo);
                List<CRN_PatientInfoVM> patList = JsonConvert.DeserializeObject<List<CRN_PatientInfoVM>>(strPatData);
                if (patList != null && patList.Count > 0)
                {
                    retObj = patList.First();
                }
            }

            return retObj;
        }

    }

    public class CRN_InvoiceInfoVM
    {
        public int PatientId { get; set; }
        public int BillingTransactionId { get; set; }
        public string InvoiceCode { get; set; }
        public int InvoiceNo { get; set; }
        public string PaymentMode { get; set; }
        public DateTime InvoiceDate { get; set; }
        public string InvoiceNoFormatted { get; set; }
        public double SubTotal { get; set; }
        public double DiscountAmount { get; set; }
        public double TaxTotal { get; set; }
        public double TotalAmount { get; set; }
        public string BillStatus { get; set; }
        public string TransactionType { get; set; }
        public string InvoiceType { get; set; }
        public bool? IsInsuranceBilling { get; set; }
        public int? InsuranceProviderId { get; set; }
        public string UserName { get; set; }
        public int SettlementId { get; set; }
        public double? CashDiscount { get; set; }

        public static CRN_InvoiceInfoVM GetSingleInvoiceInfoMappedFromDbTable(DataTable invInfo)
        {
            CRN_InvoiceInfoVM retObj = new CRN_InvoiceInfoVM();
            if (invInfo != null)
            {
                string strInvData = JsonConvert.SerializeObject(invInfo);
                List<CRN_InvoiceInfoVM> invList = JsonConvert.DeserializeObject<List<CRN_InvoiceInfoVM>>(strInvData);
                if (invList != null && invList.Count > 0)
                {
                    retObj = invList.First();
                }
            }

            return retObj;
        }
    }

}
